libgnucash/app-utils/gnc-optiondb.h

```diff
@@ -0,0 +1,214 @@
+ * Boston, MA  02110-1301,  USA       gnu@gnu.org                   *
+ *                                                                  *
+\********************************************************************/
+/** @addtogroup Engine
+    @{ */
+/** @addtogroup Options
+    GnuCash Options System for Book, Report, and Stylesheet Options.
+
+    The GnuCash Options System supports two somewhat different purposes:
+    - File properties, such as business information and whether to use automatic
+      trading accounts.
+
+    - Report options for configuring and customizing reports. Most report
+      options are user-configurable through a report options dialog but some are
+      used as a way of passing enumeration values between the scheme modules
+      that generate the report.
+
+    The options system is centered on an options database or optiondb. A
+    separate optionsdb is created and instantiated for every use, so the book
+    gets one at the beginning of the session with its values loaded from KVP,
+    and every report gets one when the report is run, as do report stylesheets
+    when they are edited. Customized report and stylesheet options are saved as
+    Scheme code fragments in files in the user's GnuCash Config directory.
+
+    @note
+    Persistence via text scheme code is a security vulnerability as it
+    enables an attacker to make GnuCash execute arbitrary code. The Guile
+    interpreter affords full system access with at least the user's privileges.
+
+    @{ */
+/** @file gnc-optiondb.h
+    @brief C public interface for the Options Database.
+    @author Copyright 2019-2021 John Ralls <jralls@ceridwen.us>
+*/
+#ifndef GNC_OPTIONDB_H_
+#define GNC_OPTIONDB_H_
+
+#ifdef __cplusplus
+class GncOption;
+class GncOptionDB;
+#else
+// It's a class in C++ but the C compiler can't tell.
+typedef struct GncOption GncOption;
+typedef struct GncOptionDB GncOptionDB;
+#endif
+
+#include <libguile.h>
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+#include <config.h>
+#include <Account.h>
+#include <gnc-budget.h>
+#include <gnc-commodity.h>
+#include <gncInvoice.h>
+#include <gncTaxTable.h>
+
+/**
+ * Create an empty option database.
+ *
+ * @return A newly allocated GncOptionDB. Use delete to destroy it.
+ */
+GncOptionDB* gnc_option_db_new(void);
+
+/**
+ * Destruct and release a GncOptionDB.
+ * @param odb The GncOptionDB.
+ */
+void gnc_option_db_destroy(GncOptionDB* odb);
+
+/**
+ * Obtain a GncOptionDB* from Scheme
+ *
+ * When report or stylesheet options are generated in Scheme the GncObjectDB is
+ * wrapped in a std::unique_ptr and then in a Guile SMOB by SWIG. The GUI code
+ * needs a reference to the GncObjectDB and we don't want to introduce swig
+ * library dependencies.
+ *
+ * @param dispatch The scheme dispatch function returned by gnc:new-options
+ * @return GncOptiondDB* Do not free this pointer!
+ */
+GncOptionDB*
+gnc_get_optiondb_from_dispatcher(SCM dispatcher);
```

Comment:
I've only noticed this here for the first time you're using SCM parameters in the public api. Is gnc-optiondb.[ch] intended to be moved into engine as well ?
That would reintroduce SCM dependencies in engine. Something I'd prefer to avoid.

Reply:
Put in another way, can the public api that depends on guile move to bindings and  does that need extra separation in the code you have currently written ?

Reply:
Hmm. It's called from window-report.cpp and dialog-report-style-sheet.c. Both are already juggling Scheme so I guess it would work to move it into gnc-optiondb.i and have those two functions call it as a scheme function `gnc-get-optiondb-from-dispatcher`.

Can you think of a better way?

Reply:
Oh, I like that much better.

Reply:
There's a worse instance for Scheme dependencies: QofQuery and internal options use GncOptionValue<SCM>. While they're at present accessed only from the Scheme API that template specialization has to be included in the GncOption variant list and that in turn makes GncOption dependent on libguile.

I think the workaround is to instead use GncOptionValue<std::string> and using `format` and `eval` in options.scm to serialize and de-serialize them.

Reply:
From what I see it boils down to the use of `gnc_make_SCM_option` in `gnc:make-query-option` (deprecated),  `gnc:make-internal-option` (deprecated) and the SCM parameter to `gnc_register_internal_option` ?

The two deprecated functions can be made to call their suggested replacements, which brings us back to C++ land only.

Looking at all instances of internal options we currently use only two real types: either strings or booleans.
Can we provide overloads for `gnc_register_internal_option` that take a std::string or  boolean as `value` parameter ? That would eliminate the need for an overload with an SCM type `value` parameter. I'm not sure though how swig will determine which function overload to use. Perhaps that needs some guidance from our side in the interface file.
If we require more internal types, we can add more overrides.

I didn't do an exhaustive scan so perhaps I have missed parts.

Reply:
Sort of. I've got Query sorted so that it correctly stores the QofQuery* and doesn't need SCM exposed in the C API. As you say, Internal options are used *mostly* for strings and bools and those can be separated in gnc:make-internal-option or the wrapper for GncOptiondb::register_internal_option. There are two exceptions: account-piecharts.scm wraps a bool in a lambda, but all of the lambdas return a fixed value so it can be trivially changed to be a bool.
The other is a bit ugly: view-column.scm writes a list of lists that I'd express in C++ as `std::vector<std::tuple<size_t, size_t, size_t>>` representing the report template and how many columns and rows it should occupy.

Reply:
> Sort of. I've got Query sorted so that it correctly stores the QofQuery* and doesn't need SCM exposed in the C API. As you say, Internal options are used _mostly_ for strings and bools and those can be separated in gnc:make-internal-option or the wrapper for GncOptiondb::register_internal_option. There are two exceptions: account-piecharts.scm wraps a bool in a lambda, but all of the lambdas return a fixed value so it can be trivially changed to be a bool.

Let's do this in maint now rather than later?

> The other is a bit ugly: view-column.scm writes a list of lists that I'd express in C++ as `std::vector<std::tuple<size_t, size_t, size_t>>` representing the report template and how many columns and rows it should occupy.

Haven't explored the C side therefore I'll avoid touching it.

Reply:
@christopherlam 
> Let's do this in maint now rather than later?

Sure.

> Haven't explored the C side therefore I'll avoid touching it.

Me either, yet. I've been avoiding the multicolumn reports, but in testing today I found that there's more option dialog stuff in gnucash/gnome/dialog-report-column-view.c and gnc-plugin-page-report.c and that it's broken in this branch so I'm going to have to dig into it soon.

Reply:
I've got the multicolumn reports working, but they're implemented in a really ugly way and need to be rewritten with its own option type that connects directly to the UI element. That will eliminate the need for the `GncValue<SCM>` variant type.

Reply:
Multicolumn reports now has its own GncValue type and all of the SCM API is removed.